Bug/invalid select value

dmbarry86 opened the pull request:
Catch the case where invalid value is passed in and just display empty in this case as the underlying select component requires. An invalid value is one that is not a member of the options.

Closes https://github.com/IPG-Automotive-UK/ngd/issues/238

mattcorner commented on src/Select/Select.test.js:
```diff
@@ -49,4 +49,18 @@ describe("Select", () => {
     const inputBase = container.querySelector(".MuiInputBase-root");
     expect(inputBase).toHaveClass("Mui-error");
   });
+  test("display empty if value is not a valid option", async () => {
```
These tests don't need to be async as they don't await anything.

mattcorner replied:
```suggestion
  test("display empty if value is not a valid option", () => {
```

mattcorner commented on src/Select/Select.test.js:
```diff
@@ -49,4 +49,18 @@ describe("Select", () => {
+    const baseComponent = container.querySelector(".MuiSelect-nativeInput");
+    expect(baseComponent.value).toBe("");
+  });
+  test("display the value if it's a valid option", async () => {
```
```suggestion
  test("display the value if it's a valid option", () => {
```

mattcorner reviewed (changes requested):
Remove async from tests and its good to go.

dmbarry86 commented:
I've removed the async from the new tests. I've also added an eslint rule to catch such cases in the future. In adding the rule I found some existing code which was violating the rule so I have fixed those as well. This then uncovered some tests which were never going to fail so I fixed those as well.

mattcorner commented on src/LoginForm/LoginForm.test.js:
```diff
@@ -24,24 +24,21 @@ describe("LoginForm", () => {
-    expect(onLogin).not.toHaveBeenCalled();
+    userEvent.type(elements.inputs.email, "joe.bloggs"); // missing password
+    fireEvent.submit(elements.submit);
+    await waitFor(() => expect(onLogin).not.toHaveBeenCalled());
```
This should now be testing for the error message on screen as we discovered. onLogin is not called by default.

mattcorner commented on src/PasswordChangeDialog/PasswordChangeDialog.test.js:
```diff
@@ -33,28 +33,26 @@ describe("PasswordChangeDialog", () => {
-    expect(onSubmit).not.toHaveBeenCalled();
+    userEvent.type(elements.inputs.newPassword, "abc123"); // top 100 password
+    fireEvent.submit(elements.submit);
+    await waitFor(() => expect(onSubmit).not.toHaveBeenCalled());
```
Test should be testing for error message on screen rather than lack of callback execution.

mattcorner commented on src/PasswordChangeForm/PasswordChangeForm.test.js:
```diff
@@ -26,43 +26,43 @@ describe("PasswordChangeForm", () => {
-    expect(onSubmit).not.toHaveBeenCalled();
+    userEvent.type(elements.inputs.password, "abc123"); // common password
+    fireEvent.submit(elements.submit);
+    await waitFor(() => expect(onSubmit).not.toHaveBeenCalled());
```
Change to test for error on screen.

mattcorner commented on src/PasswordChangeForm/PasswordChangeForm.test.js:
```diff
@@ -26,43 +26,43 @@ describe("PasswordChangeForm", () => {
+      userEvent.type(elements.inputs.password, "something");
+      await waitFor(() =>
+        expect(
+          screen.findByText("Password strength: 0/4. Minimum required 3+.")
```
should be queryByText and toBeInTheDocument()

mattcorner commented on src/PasswordChangeForm/PasswordChangeForm.test.js:
```diff
@@ -26,43 +26,43 @@ describe("PasswordChangeForm", () => {
+      userEvent.type(elements.inputs.password, "something");
+      await waitFor(() =>
+        expect(
+          screen.findByText(
```
should be queryByText and toBeInTheDocument()

mattcorner commented on src/RegistrationForm/RegistrationForm.test.js:
```diff
@@ -34,36 +34,32 @@ describe("RegistrationForm", () => {
+    // incorrect email format
+    // missing first, lastname, team, password + password repeat
+    fireEvent.submit(elements.submit);
+    await waitFor(() => expect(onRegister).not.toHaveBeenCalled());
```
Test for error message on screen instead

mattcorner reviewed (changes requested):
Some changes needed now that we've learned a bit more about the tests. I'll action as they're from me.

mattcorner reviewed (approved):
LGTM (finally)

dmbarry86 commented on src/LoginForm/LoginForm.test.js:
```diff
@@ -41,4 +41,14 @@ describe("LoginForm", () => {
   });
+  it("displays error message to user on validation fail", async () => {
+    const elements = setup();
+    userEvent.type(elements.inputs.email, "joe.bloggs"); // missing password
```
shouldn't userEvent be wrapped in act @mattcorner ?

mattcorner replied:
You'd think so, but not in this case because the form validation is only applied onSubmit, not onTouch like it is in something like RegistrationForm.

dmbarry86 replied:
Also, it's testing wrong email not missing password as the comment suggests. Password is included on this form but that's not what you are testing for.

dmbarry86 replied:
> You'd think so, but not in this case because the form validation is only applied onSubmit, not onTouch like it is in something like RegistrationForm.

So we're saying the rule is only wrap userEvent with act if the userEvent call triggers some async code from the react-hook-form library which updates the DOM? In all other cases you don't need to wrap userEvent?

dmbarry86 commented on src/PasswordResetForm/PasswordResetForm.test.js:
```diff
@@ -42,4 +42,14 @@ describe("PasswordResetForm", () => {
   });
+  it("displays error message to user on validation fail", async () => {
+    const elements = setup();
+    userEvent.type(elements.inputs.email, "joe.bloggs"); // missing password
```
shouldn't userEvent be wrapped in act @mattcorner ?

mattcorner replied:
You'd think so, but not in this case because the form validation is only applied onSubmit, not onTouch like it is in something like RegistrationForm.

dmbarry86 replied:
Also, it's testing wrong email not missing password as the comment suggests. Don't think there is a password on this form.

dmbarry86 commented on src/LoginForm/LoginForm.test.js:
```diff
@@ -24,24 +24,31 @@ describe("LoginForm", () => {
+  });
+  it("displays error message to user on validation fail", async () => {
+    const elements = setup();
+    userEvent.type(elements.inputs.email, "joe.bloggs"); // missing password
```
It's wrong email not missing password as the comment suggests
